### PR TITLE
Add support for testonly artifacts

### LINKED ISSUE #346: Add support for testonly deps
It would be nice if it was possible to specify artifacts as `testonly = True`, similar to what is possible with `neverlink = True`.

### LINKED ISSUE #364: Cannot build on Windows
On Bazel CI, at 0a7cc6a0b6764232a0ddd31ad87b489e1d47b166, with Bazel 2.0.0 (and also at HEAD: https://github.com/bazelbuild/bazel/commit/f5b9d930a93fadd40295ffe38dec2bf80569e38b) on Windows, we get build failures.

The look the same as https://github.com/bazelbuild/examples/issues/146

Log: https://buildkite.com/bazel/rules-jvm-external-examples/builds/1013#28fc946e-19ff-4c4b-86af-a34f00d670e3

```

(00:15:13) INFO: From Dexing external/maven/_dx/org_ccil_cowan_tagsoup_tagsoup/stamped_tagsoup-1.2.jar_desugared.jar with applicable dexopts []:
--
  | warning: Ignoring InnerClasses attribute for an anonymous inner class
  | (org.ccil.cowan.tagsoup.Parser$1) that doesn't come with an
  | associated EnclosingMethod attribute. This class was probably produced by a
  | compiler that did not target the modern .class file format. The recommended
  | solution is to recompile the class from source, using an up-to-date compiler
  | and without specifying any "-target" type options. The consequence of ignoring
  | this warning is that reflective operations on this class will incorrectly
  | indicate that it is *not* an inner class.
  | (00:15:13) ERROR: D:/b/hmndq6yn/external/maven/BUILD:173:1: Couldn't build file external/maven/com_android_support_appcompat_v7_symbols/symbols.zip: Compiling Android resources for @maven//:com_android_support_appcompat_v7 failed (Exit 1)
  | Jan 22, 2020 12:15:13 AM java.util.stream.ForEachOps$ForEachOp$OfRef accept
  | SEVERE: Error during Compiling bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-hdpi-v4\abc_btn_switch_to_on_mtrl_00001.9.png:
  | Command: external\androidsdk\build-tools\29.0.2\aapt2.exe\
  | compile\
  | -v\
  | --legacy\
  | -o\
  | D:\temp\android_resources_tmp15986696223201690866\compiled\bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\
  | bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-hdpi-v4\abc_btn_switch_to_on_mtrl_00001.9.png
  | Output:
  | bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-hdpi-v4\abc_btn_switch_to_on_mtrl_00001.9.png: note: compiling PNG.
  |  
  | Jan 22, 2020 12:15:13 AM java.util.stream.ForEachOps$ForEachOp$OfRef accept
  | SEVERE: Error during Compiling bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-xhdpi-v4\abc_ab_share_pack_mtrl_alpha.9.png:
  | Command: external\androidsdk\build-tools\29.0.2\aapt2.exe\
  | compile\
  | -v\
  | --legacy\
  | -o\
  | D:\temp\android_resources_tmp15986696223201690866\compiled\bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\
  | bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-xhdpi-v4\abc_ab_share_pack_mtrl_alpha.9.png
  | Output:
  | bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-xhdpi-v4\abc_ab_share_pack_mtrl_alpha.9.png: note: compiling PNG.
  |  
  | Jan 22, 2020 12:15:13 AM java.util.stream.ForEachOps$ForEachOp$OfRef accept
  | SEVERE: Error during Compiling bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-xhdpi-v4\abc_list_divider_mtrl_alpha.9.png:
  | Command: external\androidsdk\build-tools\29.0.2\aapt2.exe\
  | compile\
  | -v\
  | --legacy\
  | -o\
  | D:\temp\android_resources_tmp15986696223201690866\compiled\bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\
  | bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-xhdpi-v4\abc_list_divider_mtrl_alpha.9.png
  | Output:
  | bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-xhdpi-v4\abc_list_divider_mtrl_alpha.9.png: note: compiling PNG.
  |  
  | Jan 22, 2020 12:15:13 AM java.util.stream.ForEachOps$ForEachOp$OfRef accept
  | SEVERE: Error during Compiling bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-xhdpi-v4\abc_spinner_mtrl_am_alpha.9.png:
  | Command: external\androidsdk\build-tools\29.0.2\aapt2.exe\
  | compile\
  | -v\
  | --legacy\
  | -o\
  | D:\temp\android_resources_tmp15986696223201690866\compiled\bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\
  | bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-xhdpi-v4\abc_spinner_mtrl_am_alpha.9.png
  | Output:
  | bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-xhdpi-v4\abc_spinner_mtrl_am_alpha.9.png: note: compiling PNG.
  |  
  | Exception in thread "main" com.google.devtools.build.android.aapt2.ResourceCompiler$CompileError
  | at com.google.devtools.build.android.aapt2.ResourceCompiler$CompileError.of(ResourceCompiler.java:110)
  | at com.google.devtools.build.android.aapt2.ResourceCompiler$CompilingVisitor.getCompiledArtifacts(ResourceCompiler.java:422)
  | at com.google.devtools.build.android.aapt2.ResourceCompiler.getCompiledArtifacts(ResourceCompiler.java:456)
  | at com.google.devtools.build.android.UnvalidatedAndroidData.compile(UnvalidatedAndroidData.java:103)
  | at com.google.devtools.build.android.CompileLibraryResourcesAction.main(CompileLibraryResourcesAction.java:140)
  | at com.google.devtools.build.android.ResourceProcessorBusyBox$Tool$10.call(ResourceProcessorBusyBox.java:120)
  | at com.google.devtools.build.android.ResourceProcessorBusyBox.processRequest(ResourceProcessorBusyBox.java:237)
  | at com.google.devtools.build.android.ResourceProcessorBusyBox.main(ResourceProcessorBusyBox.java:198)
  | Suppressed: java.lang.RuntimeException: Error during Compiling bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-hdpi-v4\abc_btn_switch_to_on_mtrl_00001.9.png:
  | Command: external\androidsdk\build-tools\29.0.2\aapt2.exe\
  | compile\
  | -v\
  | --legacy\
  | -o\
  | D:\temp\android_resources_tmp15986696223201690866\compiled\bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\
  | bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-hdpi-v4\abc_btn_switch_to_on_mtrl_00001.9.png
  | Output:
  | bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-hdpi-v4\abc_btn_switch_to_on_mtrl_00001.9.png: note: compiling PNG.
  |  
  | at com.google.devtools.build.android.CommandHelper.execute(CommandHelper.java:42)
  | at com.google.devtools.build.android.AaptCommandBuilder.execute(AaptCommandBuilder.java:297)
  | at com.google.devtools.build.android.aapt2.ResourceCompiler$CompileTask.compile(ResourceCompiler.java:226)
  | at com.google.devtools.build.android.aapt2.ResourceCompiler$CompileTask.call(ResourceCompiler.java:173)
  | at com.google.devtools.build.android.aapt2.ResourceCompiler$CompileTask.call(ResourceCompiler.java:120)
  | at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
  | at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
  | at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
  | at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
  | at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
  | at java.base/java.lang.Thread.run(Thread.java:834)
  | Suppressed: java.lang.RuntimeException: Error during Compiling bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-xhdpi-v4\abc_ab_share_pack_mtrl_alpha.9.png:
  | Command: external\androidsdk\build-tools\29.0.2\aapt2.exe\
  | compile\
  | -v\
  | --legacy\
  | -o\
  | D:\temp\android_resources_tmp15986696223201690866\compiled\bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\
  | bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-xhdpi-v4\abc_ab_share_pack_mtrl_alpha.9.png
  | Output:
  | bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-xhdpi-v4\abc_ab_share_pack_mtrl_alpha.9.png: note: compiling PNG.
  |  
  | at com.google.devtools.build.android.CommandHelper.execute(CommandHelper.java:42)
  | at com.google.devtools.build.android.AaptCommandBuilder.execute(AaptCommandBuilder.java:297)
  | at com.google.devtools.build.android.aapt2.ResourceCompiler$CompileTask.compile(ResourceCompiler.java:226)
  | at com.google.devtools.build.android.aapt2.ResourceCompiler$CompileTask.call(ResourceCompiler.java:173)
  | at com.google.devtools.build.android.aapt2.ResourceCompiler$CompileTask.call(ResourceCompiler.java:120)
  | at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
  | at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
  | at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
  | at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
  | at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
  | at java.base/java.lang.Thread.run(Thread.java:834)
  | Suppressed: java.lang.RuntimeException: Error during Compiling bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-xhdpi-v4\abc_list_divider_mtrl_alpha.9.png:
  | Command: external\androidsdk\build-tools\29.0.2\aapt2.exe\
  | compile\
  | -v\
  | --legacy\
  | -o\
  | D:\temp\android_resources_tmp15986696223201690866\compiled\bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\
  | bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-xhdpi-v4\abc_list_divider_mtrl_alpha.9.png
  | Output:
  | bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-xhdpi-v4\abc_list_divider_mtrl_alpha.9.png: note: compiling PNG.
  |  
  | at com.google.devtools.build.android.CommandHelper.execute(CommandHelper.java:42)
  | at com.google.devtools.build.android.AaptCommandBuilder.execute(AaptCommandBuilder.java:297)
  | at com.google.devtools.build.android.aapt2.ResourceCompiler$CompileTask.compile(ResourceCompiler.java:226)
  | at com.google.devtools.build.android.aapt2.ResourceCompiler$CompileTask.call(ResourceCompiler.java:173)
  | at com.google.devtools.build.android.aapt2.ResourceCompiler$CompileTask.call(ResourceCompiler.java:120)
  | at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
  | at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
  | at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
  | at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
  | at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
  | at java.base/java.lang.Thread.run(Thread.java:834)
  | Suppressed: java.lang.RuntimeException: Error during Compiling bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-xhdpi-v4\abc_spinner_mtrl_am_alpha.9.png:
  | Command: external\androidsdk\build-tools\29.0.2\aapt2.exe\
  | compile\
  | -v\
  | --legacy\
  | -o\
  | D:\temp\android_resources_tmp15986696223201690866\compiled\bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\
  | bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-xhdpi-v4\abc_spinner_mtrl_am_alpha.9.png
  | Output:
  | bazel-out\android-armeabi-v7a-fastbuild\bin\external\maven\_aar\unzipped\resources\com_android_support_appcompat_v7\res\drawable-xhdpi-v4\abc_spinner_mtrl_am_alpha.9.png: note: compiling PNG.
  |  
  | at com.google.devtools.build.android.CommandHelper.execute(CommandHelper.java:42)
  | at com.google.devtools.build.android.AaptCommandBuilder.execute(AaptCommandBuilder.java:297)
  | at com.google.devtools.build.android.aapt2.ResourceCompiler$CompileTask.compile(ResourceCompiler.java:226)
  | at com.google.devtools.build.android.aapt2.ResourceCompiler$CompileTask.call(ResourceCompiler.java:173)
  | at com.google.devtools.build.android.aapt2.ResourceCompiler$CompileTask.call(ResourceCompiler.java:120)
  | at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
  | at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
  | at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
  | at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
  | at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
  | at java.base/java.lang.Thread.run(Thread.java:834)


```

### PR DESCRIPTION
**Pull request by Yannic**
Fixes #346

### CODE REVIEW

**Comment by Yannic** [comment]
Testing this is tricky because it'll fail during analysis(?).

**Comment by jin** [comment]
`diff_test` is fussy on Windows, especially with line endings. Either use a regular `sh_test` or see the tests for `manifest_stamp`: https://github.com/bazelbuild/rules_jvm_external/tree/master/tests/unit/manifest_stamp

**Comment by jin** [comment]
Could you also add a section to the main README, like `neverlink`?

File: tests/integration/BUILD [hunk 1]
```diff
@@ -1,3 +1,40 @@
+load("@bazel_skylib//rules:diff_test.bzl", "diff_test")
+
+genquery(
```
Comment (jin) [inline tests/integration/BUILD]: Very clever using `genquery`!

**Comment by Yannic** [comment]
Thanks, Readme updated :).

There is one issue when an artifact is marked as testonly but also in the transitive closure of another artifact that isn't testonly. Let's ignore this limitation for now.

**Comment by jin** [comment]
yeah, although I'll say that's a feature of `testonly` in itself :smile:

**Pull request merged**